services/src/main/java/io/scalecube/services/ServiceDispatcher.java

```diff
@@ -0,0 +1,102 @@
+package io.scalecube.services;
+
+import io.scalecube.cluster.ICluster;
+import io.scalecube.transport.Message;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+public class ServiceDispatcher {
+
+  private final ICluster cluster;
+  private final IServiceRegistry registry;
+
+  /**
+   * ServiceDispatcher constructor to listen on incoming network service request.
+   * 
+   * @param cluster instance to listen on events.
+   * @param registry service registry instance for dispatching.
+   */
+  public ServiceDispatcher(ICluster cluster, IServiceRegistry registry) {
+    this.cluster = cluster;
+    this.registry = registry;
+
+    // Start listen messages
+    cluster.listen()
+        .filter(message -> message.qualifier() != null)
+        .subscribe(this::onServiceRequest);
+  }
+
+  private void onServiceRequest(Message message) {
+    Optional<ServiceInstance> serviceInstance = 
+        registry.getLocalInstance(message.qualifier(), message.header(ServiceHeaders.METHOD));
+    
+    try {
+      if (serviceInstance.isPresent()) {
+        Object result = serviceInstance.get().invoke(message, null);
+        if (result != null) {
+          if (result instanceof Throwable) {
+            repleyWithError(message, Throwable.class.cast(result));
+          } else if (result instanceof CompletableFuture) {
+            handleComputable(cluster, message, result);
+          } else { // this is a sync request response call
+            handleSync(cluster, message, result);
+          }
+        }
+      } else {
+        repleyWithError(message,
+            new IllegalStateException("no local service instance was found for service request: [" + message + "]"));
+      }
+    } catch (Exception ex) {
+      repleyWithError(message, ex);
+    }
+
+  }
+
+  private void repleyWithError(Message message, Throwable ex) {
+    Message errorResponseMsg = Message.builder()
+        .data(ex)
```

Comment:
I mean do we really need to serialize and send through the network exception, could we send only message and Exception type and recreate it on receiver side, or we need to have remote stacktrace also?